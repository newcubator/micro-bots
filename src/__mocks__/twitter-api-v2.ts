export const TwitterApi = jest.fn();

TwitterApi.mockImplementation(() => {
  return {
    v1: {
      tweet: jest.fn().mockReturnValue({ id_str: "", full_text: "" }),
    },
    v2: {
      userByUsername: jest.fn(),
      userTimeline: jest.fn(),
    },
  };
});

export const secretsExample = {
  client_email: "email@email.com",
  private_key: "privatekey123",
};

export const getRowsExample = {
  data: {
    values: [
      ["Guid", "Title"],
      ["111111", "Wie teste ich Rust und Java"],
      ["111115", "How to do something"],
    ],
  },
};

export const fakeRssFeedItemLong = {
  creator: "Max Mustermann",
  title:
    "j4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAURj4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAURj4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAURj4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAURj4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAURj4RkRKjmHoV3iOG5d87vg4VmP9IVXvvgeznXpiGUru7WJodAUR",
  link: "FakeLink",
  pubDate: "FakeDate",
  content: "FakeContent",
  contentSnippet: "FakeSnippetContent",
  guid: "1234",
  isoDate: "FakeIsoDate",
};

export const fakeRssFeedItemShort = {
  creator: "Max Mustermann",
  title: "j4RkRKjmHoV3iR",
  link: "FakeLink",
  pubDate: "FakeDate",
  content: "FakeContent",
  contentSnippet: "FakeSnippetContent",
  guid: "1234",
  isoDate: "FakeIsoDate",
};

//export const emptyFeedArray = [{}];

export const fakeRssFeed = [
  {
    creator: "Max Mustermann",
    title: "Wie teste ich Rust und Java",
    link: "FakeLink",
    pubDate: "FakeDate",
    content: "FakeContent",
    contentSnippet: "FakeSnippetContent",
    guid: "111111",
    isoDate: "FakeIsoDate",
  },
  {
    creator: "Max Mustermann",
    title: "Wie teste ich Rust und Java 2",
    link: "FakeLink",
    pubDate: "FakeDate",
    content: "FakeContent",
    contentSnippet: "FakeSnippetContent",
    guid: "111112",
    isoDate: "FakeIsoDate",
  },
  {
    creator: "Max Mustermann",
    title: "Wie teste ich Python",
    link: "FakeLink",
    pubDate: "FakeDate",
    content: "FakeContent",
    contentSnippet: "FakeSnippetContent",
    guid: "111113",
    isoDate: "FakeIsoDate",
  },
  {
    creator: "Max Mustermann",
    title: "How to do something 3",
    link: "FakeLink",
    pubDate: "FakeDate",
    content: "FakeContent",
    contentSnippet: "FakeSnippetContent",
    guid: "111114",
    isoDate: "FakeIsoDate",
  },
  {
    creator: "Max Mustermann",
    title: "How to do something",
    link: "FakeLink",
    pubDate: "FakeDate",
    content: "FakeContent",
    contentSnippet: "FakeSnippetContent",
    guid: "111115",
    isoDate: "FakeIsoDate",
  },
];

export const fakeGoogleSheet = [
  {
    guid: "111111",
    title: "Wie teste ich Rust und Java",
  },
  {
    guid: "111115",
    title: "How to do something",
  },
];
