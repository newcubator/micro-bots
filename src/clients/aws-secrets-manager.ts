import { GetSecretValueCommandOutput, SecretsManager } from "@aws-sdk/client-secrets-manager";

export class AwsSecretsManager {
  static awsSecretsManager = new SecretsManager({
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID_GOOGLE,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY_GOOGLE,
    },
    region: "eu-central-1",
  });

  static getSecret = (secretName: string): Promise<GetSecretValueCommandOutput> => {
    return new Promise<GetSecretValueCommandOutput>((resolve, reject) => {
      this.awsSecretsManager.getSecretValue({ SecretId: secretName }, (err, data) => {
        if (err) {
          console.log(err);
          reject(err);
        } else {
          resolve(JSON.parse(data.SecretString));
        }
      });
    });
  };
}
