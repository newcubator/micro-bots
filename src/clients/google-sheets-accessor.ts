import { GaxiosResponse } from "gaxios/build/src/common";
import { google, sheets_v4 } from "googleapis";
import { GoogleAuth } from "googleapis-common";
import Schema$ValueRange = sheets_v4.Schema$ValueRange;
import Sheets = sheets_v4.Sheets;

export class GoogleSheetsAccessor {
  private googleAuth: GoogleAuth;

  private GOOGLE_CREDENTIALS;

  private client: any;

  private googleSheets: Sheets;

  public setupGoogle(credentials: any): Promise<boolean> {
    this.GOOGLE_CREDENTIALS = credentials;
    return new Promise<boolean>((resolve) => {
      this.setGoogleAuth()
        .then(() => this.authGoogle())
        .then(() => resolve(true));
    });
  }

  async getRows(spreadsheetId: string, range: string): Promise<GaxiosResponse<Schema$ValueRange>> {
    return new Promise((resolve, reject) => {
      this.googleSheets.spreadsheets.values
        .get({
          auth: this.googleAuth,
          spreadsheetId,
          range,
        })
        .then((rows) => {
          resolve(rows);
        })
        .catch((err) => {
          reject(err);
        });
    });
  }

  async addRows(spreadsheetId: string, range: string, rows: string[][]): Promise<any> {
    return this.googleSheets.spreadsheets.values.append({
      auth: this.googleAuth,
      valueInputOption: "RAW",
      spreadsheetId,
      range,
      requestBody: { values: rows, majorDimension: "COLUMNS" },
    });
  }

  private setGoogleAuth(): Promise<boolean> {
    return new Promise<boolean>((resolve) => {
      this.googleAuth = new google.auth.GoogleAuth({
        credentials: {
          client_email: this.GOOGLE_CREDENTIALS.client_email,
          private_key: this.GOOGLE_CREDENTIALS.private_key,
        },
        scopes: ["https://www.googleapis.com/auth/spreadsheets"],
      });
      resolve(true);
    });
  }

  private async authGoogle(): Promise<any> {
    return await this.googleAuth
      .getClient()
      .then((client) => {
        this.client = client;
        this.googleSheets = google.sheets({ version: "v4", auth: this.client }) as Sheets;
      })
      .catch((err) => {
        console.log(err);
      });
  }
}
