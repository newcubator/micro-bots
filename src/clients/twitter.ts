import { TwitterApi } from "twitter-api-v2";

export const twitterClient = new TwitterApi({
  appKey: process.env.TWITTER_APP_KEY,
  appSecret: process.env.TWITTER_APP_SECRET,
  accessToken: process.env.TWITTER_ACCESS_TOKEN,
  accessSecret: process.env.TWITTER_ACCESS_SECRET,
});
