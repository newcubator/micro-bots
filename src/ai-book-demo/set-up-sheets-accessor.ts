import { GoogleSheetsAccessor } from "../clients/google-sheets-accessor";

export async function setUpSheetsAccessor(): Promise<GoogleSheetsAccessor> {
  const googleSheetsAccessor = new GoogleSheetsAccessor();
  await googleSheetsAccessor.setupGoogle();
  return googleSheetsAccessor;
}
