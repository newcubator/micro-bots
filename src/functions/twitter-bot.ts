import { composeTweetFromPost } from "../twitter-bot/composeTweetFromPost";
import { getAlreadyTweetedDevSquadPosts } from "../twitter-bot/get-already-tweeted-dev-squad-posts";
import { getDevSquadPosts } from "../twitter-bot/get-dev-squad-posts";
import { filterUntweetedDevSquadPosts } from "../twitter-bot/filter-untweeted-dev-squad-posts";
import { saveTweetedPost } from "../twitter-bot/save-tweeted-post";
import { setUpSheetsAccessor } from "../twitter-bot/set-up-sheets-accessor";
import { twitterClient } from "../clients/twitter";

export const handler = async () => {
  try {
    const googleSheetsAccessor = await setUpSheetsAccessor();
    const feed = await getDevSquadPosts();
    const tweets = await getAlreadyTweetedDevSquadPosts(googleSheetsAccessor);
    const notYetTweeted = filterUntweetedDevSquadPosts(feed, tweets);
    console.info(`Found ${notYetTweeted.length} possible Tweets.`);
    if (notYetTweeted.length) {
      const plannedTweet = composeTweetFromPost(notYetTweeted[0]);
      console.info(`Planned Tweeting ${plannedTweet.guid}: ${plannedTweet.message}`);

      await saveTweetedPost(googleSheetsAccessor, plannedTweet);
      console.info("Saved new Tweet.");

      const createdTweet = await twitterClient.v1.tweet(plannedTweet.message);
      console.info(`Tweeted ${createdTweet.id_str}: ${createdTweet.full_text}`);
    }
  } catch (err) {
    console.error(err);
  }
};
