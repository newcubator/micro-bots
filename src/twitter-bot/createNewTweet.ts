import { shortenFeedTitle } from "./shorten-feed-title";
import { RssFeedItem, Tweet } from "./twitter-bot";

export function createNewTweet(feedItem: RssFeedItem): Tweet {
  let tweetText = `Neues aus dem Entwicklerteam: ${feedItem.title} ${feedItem.link} #devsquad`;
  if (tweetText.length > 280) {
    tweetText = `Neues aus dem Entwicklerteam: ${shortenFeedTitle(feedItem.title, tweetText.length)}... ${
      feedItem.link
    } #devsquad`;
  }
  return { guid: feedItem.guid, title: feedItem.title, message: tweetText };
}
