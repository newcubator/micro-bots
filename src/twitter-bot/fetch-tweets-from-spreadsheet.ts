import { GoogleSheetsAccessor } from '../clients/google-sheets-accessor';
import { Tweet } from './twitter-bot';

export async function fetchTweetsFromSpreadsheet(accessor: GoogleSheetsAccessor): Promise<Tweet[]>{
    const sheetData = (await accessor.getRows(process.env.TWITTER_BOT_SPREADSHEET_ID, "Sheet1!A:C")).data
        .values;
    return parseSheetData(sheetData);
}

function parseSheetData(sheetData: string[][]): Tweet[] {
    sheetData.shift();
    return sheetData.map((tweet) => ({
        guid: tweet[0],
        title: tweet[1],
    }));
}

