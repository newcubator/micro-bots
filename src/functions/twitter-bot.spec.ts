import Parser from "rss-parser";
import { TweetV1 } from "twitter-api-v2";
import {
  fakeGoogleSheet,
  fakeRssFeed,
  fakeRssFeedItemLong,
  fakeRssFeedItemShort,
  getRowsExample,
  secretsExample,
} from "../__mocks__/twitter-api-v2";
import { AwsSecretsManager } from "../clients/aws-secrets-manager";
import { GoogleSheetsAccessor } from "../clients/google-sheets-accessor";
import { twitterClient } from "../clients/twitter";
import { createNewTweet } from "../twitter-bot/createNewTweet";
import * as fetchRssFeed from "../twitter-bot/fetch-rss-feed";
import * as fetchTweetsFromSpreadsheet from "../twitter-bot/fetch-tweets-from-spreadsheet";
import { filterUntweetedFeed } from "../twitter-bot/filter-untweeted-feed";
import * as saveToSpreadsheet from "../twitter-bot/save-to-spreadsheet";
import * as sendTweet from "../twitter-bot/send-tweet";
import * as setUpSheetsAccessor from "../twitter-bot/set-up-sheets-accessor";
import { Tweet } from "../twitter-bot/twitter-bot";
import { handler } from "./twitter-bot";

jest.mock("rss-parser");
const ParserMock = Parser as jest.MockedClass<typeof Parser>;
jest.mock("../clients/google-sheets-accessor", () => {
  return {
    GoogleSheetsAccessor: jest.fn().mockImplementation(() => ({
      getRows: () => getRowsExample,
      setupGoogle: () => Promise.resolve(true),
    })),
  };
});

describe("TwitterBot", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should shorten tweets longer than 280 chars", () => {
    const result = createNewTweet(fakeRssFeedItemLong);
    expect(result.message).toHaveLength(280);
  });

  it("should not shorten tweets shorter than 280 chars", () => {
    const result = createNewTweet(fakeRssFeedItemShort);
    expect(result.message.length).toBeLessThanOrEqual(280);
  });

  it("should always start with 'Neues aus dem Entwicklerteam:' and end with #devsquad", () => {
    const result1 = createNewTweet(fakeRssFeedItemLong);
    const result2 = createNewTweet(fakeRssFeedItemShort);
    expect(result1.message).toContain("Neues aus dem Entwicklerteam:" && "#devsquad");
    expect(result2.message).toContain("Neues aus dem Entwicklerteam:" && "#devsquad");
  });

  it("should call Newcubator Website", async () => {
    const data = await fetchRssFeed.fetchRssFeed();
    expect(data).toEqual([]);
    expect(ParserMock.prototype.parseURL).toBeCalledWith("https://newcubator.com/devsquad/rss.xml");
  });

  it("should not filter similar Feed Item", () => {
    const result = filterUntweetedFeed(fakeRssFeed, fakeGoogleSheet);
    expect(result[0]).toBe(fakeRssFeed[1]);
  });

  it("should send a tweet", async () => {
    const message = "This is a tweet";
    await sendTweet.sendTweet(message);
    expect(twitterClient.v1.tweet).toBeCalledWith(message);
  });

  it('should log out "Tweet: something"', async () => {
    console.log = jest.fn();
    const message = "This is a tweet";
    await sendTweet.sendTweet(message);
    expect(console.log).toHaveBeenCalledWith("Tweet", "", ":", "");
  });

  it("should fetch tweets from google sheet", async () => {
    const result = await fetchTweetsFromSpreadsheet.fetchTweetsFromSpreadsheet(new GoogleSheetsAccessor());
    expect(result).toMatchSnapshot();
  });

  it("should set up Sheetsaccessor", async () => {
    AwsSecretsManager.getSecret = jest.fn().mockResolvedValue(secretsExample);
    const result = await setUpSheetsAccessor.setUpSheetsAccessor();
    expect(result).toMatchSnapshot();
  });

  //TODO: mocking in this test leads to failing all tests below

  it("should send tweet", async () => {
    jest
      .spyOn(setUpSheetsAccessor, "setUpSheetsAccessor")
      .mockResolvedValueOnce({ addRows: jest.fn() } as unknown as GoogleSheetsAccessor);
    jest.spyOn(fetchTweetsFromSpreadsheet, "fetchTweetsFromSpreadsheet").mockResolvedValue(fakeGoogleSheet);
    jest.spyOn(fetchRssFeed, "fetchRssFeed").mockResolvedValue(fakeRssFeed);
    const saveToSpreadsheetMock = jest.spyOn(saveToSpreadsheet, "saveToSpreadsheet");
    const sendTweetMock = jest.spyOn(sendTweet, "sendTweet").mockResolvedValue({} as TweetV1);
    await handler();
    expect(sendTweetMock).toHaveBeenCalled();
    expect(saveToSpreadsheetMock).toHaveBeenCalled();
  });

  it("should not send", async () => {
    jest
      .spyOn(setUpSheetsAccessor, "setUpSheetsAccessor")
      .mockResolvedValueOnce({ addRows: jest.fn() } as unknown as GoogleSheetsAccessor);
    jest.spyOn(fetchTweetsFromSpreadsheet, "fetchTweetsFromSpreadsheet").mockResolvedValue(fakeGoogleSheet);
    jest.spyOn(fetchRssFeed, "fetchRssFeed").mockResolvedValueOnce([]);
    jest.spyOn(fetchTweetsFromSpreadsheet, "fetchTweetsFromSpreadsheet").mockResolvedValueOnce([]);
    const saveToSpreadsheetMock = jest.spyOn(saveToSpreadsheet, "saveToSpreadsheet");
    const sendTweetMock = jest.spyOn(sendTweet, "sendTweet").mockResolvedValue({} as TweetV1);
    await handler();
    expect(sendTweetMock).toHaveBeenCalledTimes(0);
    expect(saveToSpreadsheetMock).toHaveBeenCalledTimes(0);
  });
});
