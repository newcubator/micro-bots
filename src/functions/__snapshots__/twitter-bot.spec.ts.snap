// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwitterBot should fetch tweets from google sheet 1`] = `
Array [
  Object {
    "guid": "111111",
    "title": "Wie teste ich Rust und Java",
  },
  Object {
    "guid": "111115",
    "title": "How to do something",
  },
]
`;

exports[`TwitterBot should set up Sheetsaccessor 1`] = `
Object {
  "getRows": [Function],
  "setupGoogle": [Function],
}
`;
